opengovsg/plumber
# PLU-132: [RFC] Support Retry-After

## Problem
We need to properly support HTTP 429 or for that matter - any failure response that contains a Retry-After header.

(It's possible for servers to maliciously add Retry-After, but that just causes us to retry a few more times <.<. I don't think its a big issue .) 

## Solution
1. Modify `processAction` to pass the actual error to `handleErrorAndThrow`.
2. In `handleErrorAndThrow`, check if error came from axios and look for Retry-After header.
3. Capture this Retry-After into a new `RetriableError` type, so that we can pass it to our BullMQ backoff function.
4. Make our backoff function return the Retry-After as the delay

### Alternative Solution
Instead of modifying `processAction` to return _yet another thing_, we could store Retry-After in execution step's `errorDetails`. But DB space is more expensive than a new param so :/

## Tests
- **TODO**: Updated & new unit tests
- Check that we respect Retry-After when retrying actions. 
- Check that successful runs don't retry, even if the server's response has a Retry-After
